modules/stochastic_tools/src/userobject/PolynomialChaos.C

```diff
@@ -0,0 +1,297 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "PolynomialChaos.h"
+
+#include "Sampler.h"
+#include "CartesianProduct.h"
+
+registerMooseObject("StochasticToolsApp", PolynomialChaos);
+
+defineLegacyParams(PolynomialChaos);
+
+InputParameters
+PolynomialChaos::validParams()
+{
+  InputParameters params = SurrogateModel::validParams();
+  params.addClassDescription("Computes and evaluates polynomial chaos surrogate model.");
+
+  params.addRequiredParam<unsigned int>("order", "Maximum polynomial order.");
+  params.addRequiredParam<std::vector<DistributionName>>(
+      "distributions", "Names of the distributions samples were taken from.");
+  params.addParam<bool>(
+      "output_mean", false, "True to output mean from computed coefficients to postprocessor.");
+  params.addParam<bool>(
+      "output_stdv",
+      false,
+      "True to output standard deviation from computed coefficients to postprocessor.");
+  params.addParam<bool>(
+      "output_skew", false, "True to output skewness from computed coefficients to postprocessor.");
+  params.addParam<bool>(
+      "output_kurt", false, "True to output kurtosis from computed coefficients to postprocessor.");
+
+  return params;
+}
+
+PolynomialChaos::PolynomialChaos(const InputParameters & parameters)
+  : SurrogateModel(parameters),
+    _order(getParam<unsigned int>("order")),
+    _quad_sampler(nullptr),
+    _tuple(generateTuple(getParam<std::vector<DistributionName>>("distributions").size(), _order)),
+    _ncoeff(_tuple.size()),
+    _coeff(_ncoeff),
+    _mean_pp(nullptr),
+    _stdv_pp(nullptr),
+    _skew_pp(nullptr),
+    _kurt_pp(nullptr)
+{
+  InputParameters params = _app.getFactory().getValidParams("Receiver");
+  if (getParam<bool>("output_mean"))
+  {
+    _fe_problem.addPostprocessor("Receiver", name() + "_mean", params);
+    _mean_pp = &_fe_problem.getPostprocessorValue(name() + "_mean");
+  }
+  if (getParam<bool>("output_stdv"))
+  {
+    _fe_problem.addPostprocessor("Receiver", name() + "_stdv", params);
+    _stdv_pp = &_fe_problem.getPostprocessorValue(name() + "_stdv");
+  }
+  if (getParam<bool>("output_skew"))
+  {
+    _fe_problem.addPostprocessor("Receiver", name() + "_skew", params);
+    _skew_pp = &_fe_problem.getPostprocessorValue(name() + "_skew");
+  }
+  if (getParam<bool>("output_kurt"))
+  {
+    _fe_problem.addPostprocessor("Receiver", name() + "_kurt", params);
+    _kurt_pp = &_fe_problem.getPostprocessorValue(name() + "_kurt");
+  }
```

Comment:
We need to make a VPP object for PolynomialChaos. It is bad practice to add objects from other objects, especially without going through the factory system.

Reply:
Fair enough. I'll switch these over to an actual postprocessor. I might need help trying to parallelize it, but I'll let you know.

Reply:
Take a look at the Statistics VPP, you should be able to follow a similar model.